### PR TITLE
refactor: find the best not the first match

### PR DESCRIPTION
**Pull request by MariaPershina**
Instead of giving the first matched intent, the engine takes the longest match (with the max amount of classifiers). If there is more than one intent with this value, the one with the biggest validity is chosen.

### CODE REVIEW

File: babble/nlp/engine.py [hunk 1]
```diff
@@ -131,28 +131,54 @@ def _load_entities(self) -> Dict[str, Dict]:
...
+        if alternatives:
+            understanding = self._get_best_match(alternatives)
+            stop = time.perf_counter()
+            log.info(
+                f"Evaluated {len(self.intents)} intents in {stop - start:0.4f} seconds"
+            )
+            return understanding
         stop = time.perf_counter()
         log.info(
             f"Evaluated {len(self.intents)} intents in {stop - start:0.4f} seconds"
```
Comment (tiaero) [inline babble/nlp/engine.py:184]: I think it is not necessary to put the performance part into the "if" branch. Just stop the time as soon as you found the match no matter if you iterate over alternatives or not.

File: babble/nlp/engine.py [hunk 2]
```diff
@@ -131,28 +131,54 @@ def _load_entities(self) -> Dict[str, Dict]:
...
+            key for key in alternatives_len.keys() if alternatives_len[key] == max_len
+        ]
+        if len(longest_alternatives) == 1:
+            return longest_alternatives[0]
+        else:
+            alternatives_validity = {
+                alternative: alternative.validity()
+                for alternative in longest_alternatives
+            }
+            log.info(
```
Comment (tiaero) [inline babble/nlp/engine.py:150]: Maybe better a debug output. Once we know that this works we don't need this info in our logs.

**Review by tiaero — COMMENTED** [review]
Looks basically good to me. I have two general questions:

1. How does this effect the performance as the algorithm now iterates over every intent which matches the basic length criteria.
2. In which case does that change improve the recognition rate? How often is this function triggered?

**Comment by MariaPershina** [comment]
> Looks basically good to me. I have two general questions:
> 
> 1. How does this effect the performance as the algorithm now iterates over every intent which matches the basic length criteria.
> 2. In which case does that change improve the recognition rate? How often is this function triggered?

1. Interestingly, there was no significant change in the performance time.
2. That change (potentially) improves recognition when an intent with a smaller amount of required classifiers can be matched, but another longer one was meant. 

In general, this function itself hasn't increased the recognition rate, as for problematic commands validity was equal, resulting in a random match. But the correct one was always in the alternatives list. Thus, as soon as validity calculation works better, RT should be improved significantly.  Of course, it's arguable if in that case, we would need this step. But we need to test it first.

**Review by tiaero — APPROVED** [review]
LGTM now

**Pull request merged**